### PR TITLE
Migrate js to ts

### PR DESCRIPTION
**Pull request by terret**
**Add**:
- typescript
- babel
- types

**Update**: 
- index.js -> .ts
- index.test.js -> .ts

**Remove**: docs(/** @params */) for functions

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,12 +38,20 @@
...
   },
   "devDependencies": {
+    "@babel/cli": "7.8.3",
+    "@babel/core": "7.8.3",
+    "@babel/preset-env": "7.8.3",
+    "@babel/preset-typescript": "7.8.3",
     "@shelf/eslint-config": "0.13.0",
+    "@types/flat": "0.0.28",
+    "@types/jest": "24.0.25",
+    "@types/node": "13.1.6",
```
Comment (KnupMan) [inline package.json]:
```suggestion
    "@types/node": "12",
```

File: package.json [hunk 2]
```diff
@@ -10,16 +10,21 @@
...
+    "prepack": "yarn build",
+    "type-check": "tsc --noEmit",
+    "type-check:watch": "npm run type-check -- --watch",
+    "build": "rm -rf lib/ && yarn build:types && babel src --out-dir lib --ignore '**/*.test.ts' --extensions '.ts'",
+    "build:types": "tsc --emitDeclarationOnly --declaration --isolatedModules false --declarationDir lib"
   },
   "files": [
-    "index.js"
+    "lib"
   ],
```
Comment (vladgolubev) [inline package.json]: See https://github.com/shelfio/fast-chunk-string/commit/ba8dab183b6c0b4798a46d54890dbb5fbf015d27, it was forgotten last time

File: package.json (hunk 1 above)
Comment (vladgolubev) [inline package.json]:
```suggestion
    "@types/node": "12",
```

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
+import {unflatten} from 'flat';
+
+interface PainlessField {
+  lang: string;
+  source: string;
+  params?: {};
```
Comment (vladgolubev) [inline src/index.ts]:
```suggestion
  params?: object;
```

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import {unflatten} from 'flat';
+
+interface PainlessField {
+  lang: string;
```
Comment (vladgolubev) [inline src/index.ts]:
```suggestion
  lang: 'painless';
```

File: wallaby.config.js [hunk 5]
```diff
@@ -1,8 +1,8 @@
 module.exports = () => {
   return {
     testFramework: 'jest',
-    files: ['package.json', 'index.js', '!index.test.js'],
-    tests: ['index.test.js'],
+    files: ['package.json', 'src/index.ts', '!src/index.test.ts'],
```
Comment (KnupMan) [inline wallaby.config.js]: Can be just `files: ['package.json', 'src/**/*.ts'],`
And for tests `['src/**/*.test.ts']`

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
+import {unflatten} from 'flat';
+
+interface PainlessField {
```
Comment (vladgolubev) [inline src/index.ts]: PainlessField => PainlessScript

File: wallaby.config.js [hunk 7]
```diff
@@ -1,8 +1,8 @@
 module.exports = () => {
```
Comment (vladgolubev) [inline wallaby.config.js]: See https://github.com/shelfio/base-dotfiles/blob/master/npm-open-source/wallaby.config.js 

There more changes needed in this file when switching to TS

**Comment by vladgolubev** [comment]
@terret what about updating CircleCI config? See https://github.com/shelfio/base-dotfiles/blob/master/npm-open-source/.circleci/config.yml

File: wallaby.config.js (hunk 7 above)
Comment (terret) [reply wallaby.config.js]: Done

**Review by KnupMan — APPROVED** [review]

**Comment by terret** [comment]
> @terret what about updating CircleCI config? See [shelfio/base-dotfiles:npm-open-source/.circleci/config.yml@`master`](https://github.com/shelfio/base-dotfiles/blob/master/npm-open-source/.circleci/config.yml) [](/shelfio/base-dotfiles/blob/HEAD@{2020-01-14T16:19:31Z}/npm-open-source/.circleci/config.yml)

File: package.json [hunk 8]
```diff
@@ -50,15 +40,40 @@
...
+    "@babel/cli": "7.8.3",
+    "@babel/core": "7.8.3",
+    "@babel/preset-env": "7.8.3",
+    "@babel/preset-typescript": "7.8.3",
+    "@shelf/eslint-config": "0.13.0",
+    "@types/flat": "0.0.28",
+    "@types/jest": "24.0.25",
+    "@types/node": "12",
+    "eslint": "6.8.0",
+    "husky": "3.1.0",
```
Comment (maaraanas) [inline package.json]:
```suggestion
    "husky": "4.0.7",
```

**Review by vladholubiev — APPROVED** [review]
👍

**Pull request merged**